src/ucx_van.h

```diff
@@ -441,195 +526,332 @@ class UCXVan : public Van {
   }
 
-  uint64_t DecodeKey(SArray<char> keys) {
-    // Just one key is supported now
-    CHECK_EQ(keys.size(), 8) << "Wrong key size " << keys.size();
-    return *(reinterpret_cast<uint64_t*>(keys.data()));
+  // UCX Callbacks
+  static void ConnHandlerCb(ucp_conn_request_h conn_request, void *arg) {
+    UCXContext *ctx = reinterpret_cast<UCXContext*>(arg);
+    ctx->ep_pool_.Create(conn_request);
   }
 
-  char *GetRxBuffer(uint64_t key, int node_id, size_t size, bool push) {
-    if (!push) {
-      // Must be receive for pulled data, get the address used for push
-      std::lock_guard<std::mutex> lock(w_pool_mtx_);
-      auto addr = w_pool_.find(key);
-      CHECK(addr != w_pool_.end());
-      return addr->second;
+  static void RxMetaCompletedCb(void *request, ucs_status_t status,
+                              ucp_tag_recv_info_t *info)
+  {
+    UCXRequest *req = reinterpret_cast<UCXRequest*>(request);
+
+    CHECK_STATUS(status) << "RxMetaCompletedCb failed with " << ucs_status_string(status);
+    req->completed = true;
+    if (req->data.raw_meta == nullptr) {
+      // immediate completion
+      return;
     }
+    req->ctx->PostRecvData(req);
+  }
 
-    if (rpool_.find(key) != rpool_.end()) {
-      auto it = rpool_[key].find(node_id);
-      if (it != rpool_[key].end()) {
-        if (size <= it->second.second) {
-          return it->second.first;
-        } else {
-          // cached buffer is smaller than requested - free it and reallocate
-          free(it->second.first);
+  static void RxDataCompletedCb(void *request, ucs_status_t status,
+                              ucp_tag_recv_info_t *info)
+  {
+    UCXRequest *req = reinterpret_cast<UCXRequest*>(request);
+
+    CHECK_STATUS(status) << "RxDataCompletedCb failed with " << ucs_status_string(status);
+    req->completed = true;
+    if (req->data.buffer == nullptr) {
+      // immediate completion
+      return;
+    }
+    req->ctx->rx_pool_->Push(req->data);
+    UCX_REQUEST_FREE(req); // can release request back to UCX now
+  }
+
+  static void TxReqCompletedCb(void *request, ucs_status_t status)
+  {
+    UCXRequest *req = reinterpret_cast<UCXRequest*>(request);
+
+    CHECK_STATUS(status) << "TX request completed with " << ucs_status_string(status);
+
+    delete [] req->data.raw_meta;
+
+    UCX_REQUEST_FREE(req);
+  }
+
+  UCXEndpointsPool                                  ep_pool_;
+  ucp_context_h                                     context_;
+  ucp_worker_h                                      worker_;
+  ucp_listener_h                                    listener_;
+  int                                               src_dev_idx_;
+  Node                                              *my_node_;
+  UCXVan                                            *van_;
+  UCXRecvPool                                       *rx_pool_;
+};
+
+class UCXVan : public Van {
+ public:
+  UCXVan() {
+    setenv("UCX_USE_MT_MUTEX", "y", 0);
+    setenv("UCX_IB_NUM_PATHS", "2", 0);
```

Comment:
what's the difference between this and `UCX_MAX_RNDV_RAILS`?

Reply:
this one allows creating multiple endpoints for a single NIC

Reply:
do you mean `UCX_MAX_RNDV_RAILS ` controls the total number of QPs, and `UCX_IB_NUM_PATHS` controls the number of QPs per NIC? I remember you said `UCX_MAX_RNDV_RAILS` also controls the number of verbs QPs.

Reply:
@pleasantrabbit, yes, we can say so. Below are some more clarifications: 
`UCX_MAX_RNDV_RAILS` controls the number of network transports which can be used for single message transfer. If we use just IB transport then:
1. `UCX_MAX_RNDV_RAILS` control the total number of QPs (but using just one QP per NIC by default)
2. `UCX_IB_NUM_PATHS` controls the number of QPs per NIC (can't be more that `UCX_MAX_RNDV_RAILS`)